pkg/arvo/lib/string.hoon

```diff
@@ -0,0 +1,561 @@
+::
+::  ++octdigits
+::
+::  Eight octal digits. 
+::
+::  Source: 
+++  octdigits       ^~  `tape`(gulf 48 55)
+::
+::  ++punctuation     tape:(weld (gulf 32 47) (gulf 58 64) (gulf 91 96) (gulf 123 126))
+::
+::  All ASCII punctuation characters. 
+::
+::  Source: 
+++  punctuation     ^~  `tape`:(weld (gulf 32 47) (gulf 58 64) (gulf 91 96) (gulf 123 126))
+::
+::  ++whitespace      tape:(weld " " ~['a'] ~['9'])
+::
+::  All ASCII whitespace characters. 
+::
+::  Source: 
+++  whitespace      ^~  `tape`:(weld " " ~['a'] ~['9'])
+::
+::  ++ascii
+::
+::  All printable ASCII characters (not in ASCII order). 
+::
+::  Source: 
+++  ascii           ^~  `tape`:(weld alphabet digits punctuation whitespace)
+::
+::    Utility sets
+::
+::  ++set-alphabet
+::
+::  26-letter Roman alphabet, upper case and lower case. 
+::
+::  Source: 
+++  set-alphabet      ^~  (~(gas in *(set @tD)) alphabet)
+::
+::  ++set-alpha-lower
+::
+::  26-letter Roman alphabet, lower case. 
+::
+::  Source: 
+++  set-alpha-lower   ^~  (~(gas in *(set @tD)) alpha-lower)
+::
+::  ++set-alpha-upper
+::
+::  26-letter Roman alphabet, lower case. 
+::
+::  Source: 
+++  set-alpha-upper   ^~  (~(gas in *(set @tD)) alpha-upper)
+::
+::  ++set-digits
+::
+::  Ten decimal digits. 
+::
+::  Source: 
+++  set-digits        ^~  (~(gas in *(set @tD)) digits)
+::
+::  ++set-alpha-digits
+::
+::  26-letter Roman alphabet, upper case and lower case, plus ten decimal digits. 
+::
+::  Source: 
+++  set-alpha-digits  ^~  (~(gas in *(set @tD)) alpha-digits)
+::
+::  ++set-hexdigits
+::
+::
+::  Sixteen hexadecimal digits, upper case and lower case. 
+::
+::  Source: 
+++  set-hexdigits     ^~  (~(gas in *(set @tD)) hexdigits)
+::
+::  ++set-octdigits
+::
+::  Eight octal digits. 
+::
+::  Source: 
+++  set-octdigits     ^~  (~(gas in *(set @tD)) octdigits)
+::
+::  ++set-punctuation
+::
+::  Standard ASCII punctuation characters. 
+::
+::  Source: 
+++  set-punctuation   ^~  (~(gas in *(set @tD)) punctuation)
+::
+::  ++set-whitespace
+::
+::  Standard ASCII whitespace characters (space, newline, horizontal tab). 
+::
+::  Source: 
+++  set-whitespace    ^~  (~(gas in *(set @tD)) whitespace)
+::
+::  ++set-ascii
+::
+::  All printable ASCII characters. 
+::
+::  Source: 
+++  set-ascii         ^~  (~(gas in *(set @tD)) ascii)
+::
+::  ++is-alpha
+::
+::  Tests whether all characters in a tape are from the ASCII alphabet.
+::
+::  Source:
+++  is-alpha    |=(=tape =(~ (~(dif in (~(gas in *(set @tD)) tape)) set-alphabet)))
+::
+::  ++is-lower
+::
+::  Tests whether all characters in a tape are from the lower case ASCII alphabet.
+::
+::  Source:
+++  is-lower    |=(=tape =(~ (~(dif in (~(gas in *(set @tD)) tape)) set-alpha-lower)))
+::
+::  ++is-upper
+::
+::  Tests whether all characters in a tape are from the upper case ASCII alphabet.
+::
+::  Source: 
+++  is-upper    |=(=tape =(~ (~(dif in (~(gas in *(set @tD)) tape)) set-alpha-upper)))
```

Comment:
These functions would be easier to read if they used a helper that encapsulated checking all the characters of a tape against a set of characters, like this:
```hoon
++  is-tape-in-set
  |=  chars=(set @tD)
  |=  =tape
  ^-  ?
  =(~ (~(dif in (~(gas in *(set @tD)) tape)) chars)) 
::
++  is-upper  (is-tape-in-set set-alpha-upper)
++  is-digit  (is-tape-in-set set-digits)
```

Reply:
feel free to rename that helper function